src/Modules/Content/Dnn.PersonaBar.Pages/Components/PagesControllerImpl.cs

```diff
@@ -1224,67 +1224,66 @@ private void CopySourceTabProperties(TabInfo tab, TabInfo sourceTab)
             }
         }
 
-        private void CopyModulesFromSourceTab(TabInfo tab, TabInfo sourceTab)
+        private void CopyModulesFromSourceTab(TabInfo tab, TabInfo sourceTab, IEnumerable<ModuleItem> includedModules)
         {
-            foreach (var module in sourceTab.ChildModules.Values)
+            foreach (var module in includedModules)
             {
-                if (module.IsDeleted || module.AllTabs)
+                var includedInCopy = module.IncludedInCopy ?? true;
+                if (!includedInCopy || !sourceTab.ChildModules.Values.Any(m => !m.IsDeleted && !m.AllTabs && m.ModuleID == module.Id))
                 {
                     continue;
                 }
 
-                var newModule = module.Clone();
-
-                newModule.TabID = tab.TabID;
-                newModule.DefaultLanguageGuid = Null.NullGuid;
-                newModule.CultureCode = tab.CultureCode;
-                newModule.VersionGuid = Guid.NewGuid();
-                newModule.LocalizedVersionGuid = Guid.NewGuid();
-
-                newModule.ModuleID = Null.NullInteger;
-                _moduleController.InitialModulePermission(newModule, newModule.TabID, 0);
-                newModule.InheritViewPermissions = module.InheritViewPermissions;
-
-                newModule.ModuleID = _moduleController.AddModule(newModule);
-
-                //Copy each setting to the new TabModule instance
-                foreach (DictionaryEntry setting in module.ModuleSettings)
+                var copyType = module.CopyType ?? ModuleCopyType.Copy;
+                var objModule = ModuleController.Instance.GetModule(module.Id, sourceTab.TabID, false);
+                ModuleInfo newModule = null;
+                if (objModule != null)
                 {
-                    _moduleController.UpdateModuleSetting(newModule.ModuleID, Convert.ToString(setting.Key), Convert.ToString(setting.Value));
-                }
+                    //Clone module as it exists in the cache and changes we make will update the cached object
+                    newModule = objModule.Clone();
+                    newModule.TabID = tab.TabID;
+                    newModule.DefaultLanguageGuid = Null.NullGuid;
+                    newModule.CultureCode = tab.CultureCode;
+                    newModule.ModuleTitle = module.Title;
+
+                    if (copyType != ModuleCopyType.Reference)
+                    {
+                        newModule.ModuleID = Null.NullInteger;
+                        ModuleController.Instance.InitialModulePermission(newModule, newModule.TabID, 0);
+                    }
```

Comment:
I think we need to have the InheritViewPermissions copied as well as the Module Permissions. Please check ManageTabs.ascx.cs:1030 from 8.0.4 release branch of Platform